### PR TITLE
```Extension``` styles

### PR DESCRIPTION
**Pull request by littlewhitecloud**
**Flags**
- [x] Add ```write_style()``` and ```load_style()```
- [x] Add a gui
- [x] Improve the basic style
- [x] Improve the gui
~~Add more built-in style~~

**Tests**
/

**Note**
@Moosems Would you like to build some built-in styles? Or take part in writing config ui?
(Also, are you good at json? I need some help)

### CODE REVIEW

**Comment by littlewhitecloud** [comment]
@Moosems now you can custom your own styles (but still need to improve)

**Comment by littlewhitecloud** [comment]
@Moosems Now you can custom your own style! Would like to take part in this pr to make more built-in styles?
(By the way, are you good at json?)

**Comment by littlewhitecloud** [comment]
image
Not bad LOL

**Comment by littlewhitecloud** [comment]
Are you interested in checking this out? @Moosems

**Comment by Moosems** [comment]
After getting my sleep, sure.

**Comment by littlewhitecloud** [comment]
> 

Thank you!

**Comment by Moosems** [comment]
Anytime my friend 😁

**Comment by littlewhitecloud** [comment]
Have a good rest and I will improve it.

**Comment by Moosems** [comment]
Scroll bar:

row = 1 if horizontal else 0
col = 0 if horizontal else 1
self.grid(row=row, column=col)

File: tktermwidget/style.py [hunk 1]
```diff
@@ -1,27 +1,275 @@
...
+        pass
+    else:
+        theme = True
+        set_theme("dark")
+        config.option_add("*font", ("Cascadia Mono", 9))
+
+    style: dict[str, str, str, str, str] = DEFAULT if load_style() == "{}" else load_style()
+
+    create = Label(config, text="Create your custom style!")
+    """
```
Comment (Moosems) [inline tktermwidget/style.py:102]: What is this?

File: tktermwidget/style.py [hunk 2]
```diff
@@ -1,27 +1,275 @@
...
+# Load style
+def load_style() -> dict:
+    """Get style from the json file"""
+    # User also can use this function to get their own style
+    with open(JSON_FILE, "r", encoding="utf-8") as json_obj:
+        return load(json_obj)
+
+
+# Config style
+def config_style() -> None:
```
Comment (Moosems) [inline tktermwidget/style.py:42]: Make this into an app class instead to remove nested functions

File: tktermwidget/style.py [hunk 3]
```diff
@@ -1,27 +1,275 @@
...
+# {yourstylename}: dict[str, str, str, str, str] = {
+#    "background": "{yourhexcolor}",
+#    "insertbackground": "{yourhexcolor}",
+#    "selectbackground": "{yourhexcolor}",
+#    "selectforeground": "{yourhexcolor}",
+#    "foreground": "{yourhexcolor}",
+# }
+
+# Built-in styles
+DEFAULT: dict[str, str, str, str, str] = {  # Style for normal tkterminalwidget
```
Comment (Moosems) [inline tktermwidget/style.py:241]: Move constants to top

File: tktermwidget/style.py [hunk 4]
```diff
@@ -1,27 +1,275 @@
...
     "background": "#2B2B2B",
     "insertbackground": "#DCDCDC",
     "selectbackground": "#b4b3b3",
     "selectforeground": "#e6e6e6",
     "foreground": "#cccccc",
 }
 
-Powershell: dict = {  # Style for powershell
+POWERSHELL: dict[str, str, str, str, str] = {  # Style for powershell
```
Comment (Moosems) [inline tktermwidget/style.py:249]: Just use dict[str]. The repetition isn't quite necessary ;).

File: tktermwidget/style.py (hunk 1 above)
Comment (littlewhitecloud) [reply tktermwidget/style.py:102]: It is load style.
If there exists a user custom style, then load it, else use default style instead !

File: tktermwidget/style.py (hunk 2 above)
Comment (littlewhitecloud) [reply tktermwidget/style.py:42]: Okay, I will do it.

File: .editorconfig [hunk 5]
```diff
@@ -4,17 +4,24 @@ root = true
 charset = utf-8
 insert_final_newline = true
 
-# Special file
+# Special files
+[tktermwidget/__init__.py]
```
[the remark this answers is not in the document]
Comment (littlewhitecloud) [reply .editorconfig]: okay

File: tktermwidget/style.py (hunk 3 above)
Comment (littlewhitecloud) [reply tktermwidget/style.py]: fixed

File: tktermwidget/style.py (hunk 4 above)
Comment (littlewhitecloud) [reply tktermwidget/style.py]: Thanks for you tip! (I never wrote it before)

**Comment by littlewhitecloud** [comment]
Look strange after using  the code. 

image

File: tktermwidget/style.py (hunk 2 above)
Comment (littlewhitecloud) [reply tktermwidget/style.py]: I rebuilded it and wrote a lot
see ```rebuild style.py``` commit

**Comment by littlewhitecloud** [comment]
image
Look more beautiful now.

**Comment by littlewhitecloud** [comment]
@Moosems Can help figure out why these buttons didn't use the correct font after I do 
```python
window.option_add("*font", ("Cascadia Mono", 9, "normal))```
![image](https://github.com/littlewhitecloud/TkTerminal/assets/71159641/07e0379f-6078-410c-9a87-d4d2a769931b)

**Comment by Moosems** [comment]
None of the items have the font because you are using TTK widgets. You need to use Style().theme_configure("TButton", "font", font).

**Comment by littlewhitecloud** [comment]
> None of the items have the font because you are using TTK widgets. You need to use Style().theme_configure("TButton", "font", font).

But there are only two buttons don't have the correct font.

**Comment by littlewhitecloud** [comment]
> None of the items have the font because you are using TTK widgets. You need to use Style().theme_configure("TButton", "font", font).

Can you make a suggestion?
```console
AttributeError: 'Style' object has no attribute 'theme_configure'

**Comment by littlewhitecloud** [comment]
yes

**Comment by littlewhitecloud** [comment]
> The others are images do the font doesn't affect them...

I didn't mean the buttons with image don't follow the font, I mean the buttons in the red box don't follow font

**Comment by littlewhitecloud** [comment]
> Are you using sv_ttk?

I opened an issues on the sv_ttk github repo.

**Comment by littlewhitecloud** [comment]
nevermind about it, @Moosems do you want to add some built-in styles?

**Comment by Moosems** [comment]
Possibly, I don't personally see myself using them that much though, I'd much rather prefer to just configure the style of the terminal.

**Comment by littlewhitecloud** [comment]
Okay, it is just a little extension

File: tktermwidget/style.py [hunk 6]
```diff
@@ -1,27 +1,260 @@
...
+            command=lambda: self.selectcolor(selectforegroundentry, "selectforeground"),
+        )
+
+        foregroundframe = Frame(self)
+        foreground = Label(foregroundframe, text="Choose or input your selectforeground hex color")
+        foregroundentry = Entry(foregroundframe)
+        foregroundbutton = Button(foregroundframe, command=lambda: self.selectcolor(foregroundentry, "foreground"))
+
+        buttonframe = Frame(self)
+        buttonframe.config(background="#2a2a2a")
```
Comment (littlewhitecloud) [inline tktermwidget/style.py:149]:
```suggestion

```

File: tktermwidget/style.py [hunk 7]
```diff
@@ -1,27 +1,260 @@
...
+        save = Button(buttonframe, text="Save", width=5, command=self.savestyle)
```
Comment (littlewhitecloud) [inline tktermwidget/style.py:150]:
```suggestion
        save = Button(buttonframe, text="Save", width=6, command=self.savestyle)
```

File: tktermwidget/style.py [hunk 8]
```diff
@@ -1,27 +1,260 @@
...
+        cancel = Button(buttonframe, text="Cancel", width=5, command=self.destroy)
```
Comment (littlewhitecloud) [inline tktermwidget/style.py:151]:
```suggestion
        cancel = Button(buttonframe, text="Cancel", width=6, command=self.destroy)
```

**Comment by littlewhitecloud** [comment]
I think it is almost finish.
All todos are finished.
Need to review it again and do some bug fix on this.

**Comment by littlewhitecloud** [comment]
And then, let's go to bug fix!

**Comment by littlewhitecloud** [comment]
I am going to sleep lol. Tomorrow I have to go to school and then I will have summer holidays, at that time I will review it.

**Pull request merged**